crates/bevy_pbr/src/meshlet/visibility_buffer_resolve.wgsl

```diff
@@ -0,0 +1,184 @@
+#define_import_path bevy_pbr::meshlet_visibility_buffer_resolve
+
+#import bevy_pbr::{
+    meshlet_bindings::{
+        meshlet_visibility_buffer,
+        meshlet_thread_meshlet_ids,
+        meshlets,
+        meshlet_vertex_ids,
+        meshlet_vertex_data,
+        meshlet_thread_instance_ids,
+        meshlet_instance_uniforms,
+        get_meshlet_index,
+        unpack_meshlet_vertex,
+    },
+    mesh_view_bindings::view,
+    mesh_functions::mesh_position_local_to_world,
+    mesh_types::MESH_FLAGS_SIGN_DETERMINANT_MODEL_3X3_BIT,
+    view_transformations::{position_world_to_clip, frag_coord_to_ndc},
+}
+#import bevy_render::maths::{affine_to_square, mat2x4_f32_to_mat3x3_unpack}
+
+#ifdef PREPASS_FRAGMENT
+#ifdef MOTION_VECTOR_PREPASS
+#import bevy_pbr::{
+    prepass_bindings::previous_view_proj,
+    pbr_prepass_functions::calculate_motion_vector,
+}
+#endif
+#endif
+
+/// Functions to be used by materials for reading from a meshlet visibility buffer texture.
+
+struct PartialDerivatives {
+    barycentrics: vec3<f32>,
+    ddx: vec3<f32>,
+    ddy: vec3<f32>,
+}
+
+// https://github.com/ConfettiFX/The-Forge/blob/2d453f376ef278f66f97cbaf36c0d12e4361e275/Examples_3/Visibility_Buffer/src/Shaders/FSL/visibilityBuffer_shade.frag.fsl#L83-L139
+fn compute_partial_derivatives(vertex_clip_positions: array<vec4<f32>, 3>, ndc_uv: vec2<f32>, screen_size: vec2<f32>) -> PartialDerivatives {
+    var result: PartialDerivatives;
+
+    let inv_w = 1.0 / vec3(vertex_clip_positions[0].w, vertex_clip_positions[1].w, vertex_clip_positions[2].w);
+    let ndc_0 = vertex_clip_positions[0].xy * inv_w[0];
+    let ndc_1 = vertex_clip_positions[1].xy * inv_w[1];
+    let ndc_2 = vertex_clip_positions[2].xy * inv_w[2];
+
+    let inv_det = 1.0 / determinant(mat2x2(ndc_2 - ndc_1, ndc_0 - ndc_1));
+    result.ddx = vec3(ndc_1.y - ndc_2.y, ndc_2.y - ndc_0.y, ndc_0.y - ndc_1.y) * inv_det * inv_w;
+    result.ddy = vec3(ndc_2.x - ndc_1.x, ndc_0.x - ndc_2.x, ndc_1.x - ndc_0.x) * inv_det * inv_w;
+
+    var ddx_sum = dot(result.ddx, vec3(1.0));
+    var ddy_sum = dot(result.ddy, vec3(1.0));
+
+    let delta_v = ndc_uv - ndc_0;
+    let interp_inv_w = inv_w.x + delta_v.x * ddx_sum + delta_v.y * ddy_sum;
+    let interp_w = 1.0 / interp_inv_w;
+
+    result.barycentrics = vec3(
+        interp_w * (delta_v.x * result.ddx.x + delta_v.y * result.ddy.x + inv_w.x),
+        interp_w * (delta_v.x * result.ddx.y + delta_v.y * result.ddy.y),
+        interp_w * (delta_v.x * result.ddx.z + delta_v.y * result.ddy.z),
+    );
+
+    result.ddx *= 2.0 / screen_size.x;
+    result.ddy *= 2.0 / screen_size.y;
+    ddx_sum *= 2.0 / screen_size.x;
+    ddy_sum *= 2.0 / screen_size.y;
```

Comment:
do we know for sure the optimizer does common term extraction here and saves the couple divs

Reply:
It should (compilers ultimately shell out to LLVM usually), but I can do it just in case.